ecommerce/views.py

```diff
@@ -267,29 +279,64 @@ def _update_items(basket, items):
         items (list of JSON objects): Basket items to update, or clear if empty list, or leave as is if None
 
     Returns:
-        ProductVersion: ProductVersion object to assign to basket, if any.
+        tuple: ProductVersion object to assign to basket, if any, and a list of CourseRun
 
     """
     if items:
         if len(items) > 1:
             raise ValidationError("Basket cannot contain more than one item")
         # Item updated
-        product_version_id = items[0].get("id")
+        item = items[0]
+        product_version_id = item.get("id")
+        run_ids = item.get("run_ids")
+
         if product_version_id is None:
             raise ValidationError("Invalid request")
         try:
             product_version = ProductVersion.objects.get(id=product_version_id)
         except ProductVersion.DoesNotExist:
-            raise ValidationError(
-                "Invalid product version id {}".format(items[0]["id"])
-            )
+            raise ValidationError(f"Invalid product version id {product_version_id}")
+
+        if run_ids is not None:
+            content_object = product_version.product.content_object
+            runs_for_product = CourseRun.objects.filter(id__in=run_ids)
+            if isinstance(content_object, Course):
+                runs_for_product = runs_for_product.filter(course=content_object)
+            elif isinstance(content_object, Program):
+                runs_for_product = runs_for_product.filter(
+                    course__program=content_object
+                )
+            else:
+                raise ValidationError(
+                    f"Unknown content_object for {product_version.product}"
+                )
+
+            runs_for_product_lookup = {run.id: run for run in runs_for_product}
```

Comment:
You could probably simplify the next 6 lines or so to:

```python
run_ids_for_product = {run.id for run in runs_for_product}

missing_run_ids = set(run_id) - run_ids_for_product
if missing_run_ids:
    raise ValidationError(f"Unable to find run(s) with id(s) {missing_run_ids}")
```

Reply:
:+1: